Reorg runtime tests (3rd generation); mostly to use descriptor files for tests not annotations in *Descriptor.java files (supports later versions of java)

parrt opened the pull request:
The goal is to remove dependence on tools.jar for annotation processing.

KvanTTT commented on runtime-testsuite/test/org/antlr/v4/test/runtime/BaseRuntimeTest.java:
```diff
@@ -354,9 +356,200 @@ public static ErrorQueue antlrOnString(String workdir,
+				String gname = g.a;
+				String grammar = g.b;
+				String filename = d.getTestName()+".txt";
+				String content = "";
```
Please use `StringBuilder` for code with a big amount of concatenation. An ordinary string is bad for performance here.

ericvergnaud replied:
@KvanTTT FYI the java compiler replaces string concatenation by StringBuilder calls, so no need to change anything

KvanTTT replied:
The compiler uses allocations of `StringBuilder` for every line, it's inefficient. Maybe it works fine for dynamic languages such as JavaScript (https://stackoverflow.com/a/16696775/1046374), Python (https://stackoverflow.com/a/12171382/1046374), but not for static-like C#, Java, C++.

To not be unfounded, I've created and compiled a simple test:

```java
public static String method(String x) {
    String content = "";
    content += "1";
    content += x;
    return content;
}
```

Java <= 8 compile the code in the following way:

```
  public static java.lang.String method(java.lang.String);
    Code:
       0: ldc           #2                  // String
       2: astore_1
       3: new           #3                  // class java/lang/StringBuilder
       6: dup
       7: invokespecial #4                  // Method java/lang/StringBuilder."<init>":()V
      10: aload_1
      11: invokevirtual #5                  // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      14: ldc           #6                  // String 1
      16: invokevirtual #5                  // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      19: invokevirtual #7                  // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
      22: astore_1
      23: new           #3                  // class java/lang/StringBuilder
      26: dup
      27: invokespecial #4                  // Method java/lang/StringBuilder."<init>":()V
      30: aload_1
      31: invokevirtual #5                  // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      34: aload_0
      35: invokevirtual #5                  // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      38: invokevirtual #7                  // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
      41: astore_1
      42: aload_1
      43: areturn
```

It uses `StringBuilder` but it allocates a new instance for every line. It's inefficient.

Java >= 11 compile the code in a more effective way using `makeConcatWithConstants`:

```
  public static java.lang.String method(java.lang.String);
    Code:
       0: ldc           #2                  // String
       2: astore_1
       3: aload_1
       4: invokedynamic #3,  0              // InvokeDynamic #0:makeConcatWithConstants:(Ljava/lang/String;)Ljava/lang/String;
       9: astore_1
      10: aload_1
      11: aload_0
      12: invokedynamic #4,  0              // InvokeDynamic #1:makeConcatWithConstants:(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
      17: astore_1
      18: aload_1
      19: areturn
```

But anyway `makeConcatWithConstants` creates an allocation on every line. It's not so efficient for long concatenation especially if loops are used.

Also, `StringBuilder` is more readable in my opinion and it's more canonical for Java.

ericvergnaud replied:
That's a different pattern from:

```
public static String method(String x) {
    String content = ""
         += "1"
         + = x;
    return content;
}
```

which should be used, rather than StringBuilder

KvanTTT replied:
It differs from the original pattern with `+=` operator.

Consider the following code:

```java
public static String methodPlusAssign(String x, String y) {
    String content = "";
    content += x;
    content += y;
    return content;
}
public static String methodAssign(String x, String y) {
    String content = "" + x + y;
    return content;
}
```

And their JVM bytecode (Java 11):

```
  public static java.lang.String methodPlusAssign(java.lang.String, java.lang.String);
    Code:
       0: ldc           #2                  // String
       2: astore_2
       3: aload_2
       4: aload_0
       5: invokedynamic #3,  0              // InvokeDynamic #0:makeConcatWithConstants:(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
      10: astore_2
      11: aload_2
      12: aload_1
      13: invokedynamic #3,  0              // InvokeDynamic #0:makeConcatWithConstants:(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
      18: astore_2
      19: aload_2
      20: areturn

  public static java.lang.String methodAssign(java.lang.String, java.lang.String);
    Code:
       0: aload_0
       1: aload_1
       2: invokedynamic #3,  0              // InvokeDynamic #0:makeConcatWithConstants:(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
       7: astore_2
       8: aload_2
       9: areturn
```

There are two invocations in `methodPlusAssign` and only one in `methodPlus`. Also, using `+` is limited and such optimization does not work for loops.

parrt replied:
Hi guys. Thank you for the excellent analysis, however, I don't think performance is critical here as we load the descriptor files. The difference between 10 ms and 50 ms is not really important for this testing rig.

@ericvergnaud what do you think of the descriptor mechanism so far?

```
[grammar]
grammar T;
a : ID+ {
<writeln("$text")>
};
ID : 'a'..'z'+;
WS : (' '|'\n') -> skip;

[grammarName]
T

[start]
a

[input]
a b c

[output]
"""abc
"""
```

ericvergnaud replied:
I like it, LGTM.
Missing [slaveGrammars] ?
How about adding test name and base class name, such that we can generate test code from the test suite ?

parrt replied:
BTW I'm off on vacation 15th - 19th december.

KvanTTT commented:
There is something weird with Go runtime tests, they are hanging for now.

parrt commented:
Yeah. Can’t figure that out but it must be a small issue with input string
or something that’s causing an infinite Loop

On Wed, Dec 15, 2021 at 2:46 AM Ivan Kochurkin ***@***.***>
wrote:

> There is something weird with Go runtime tests, they are hanging for now.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/antlr/antlr4/pull/3407#issuecomment-994667414>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABLUWPOE5YO5OPGADMYO5LURBW2HANCNFSM5J7RO2XQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
-- 
Dictation in use. Please excuse homophones, malapropisms, and nonsense.

KvanTTT commented:
`.txt` is the file format for text files, but we have descriptions. Maybe replace `.txt` with `.desc` or something else?

parrt commented:
Easier to open files and .descr doesn't add much :)

I think I might have this kinda working!  I also split cpp so it finishes w/o timing out.

KvanTTT commented on runtime-testsuite/OLD-descriptors/ParserErrorsDescriptors.java:
```diff
@@ -639,7 +639,7 @@
 		@Override
 		public boolean ignore(String targetName) {
-			return !"Java".equals(targetName) && !"Swift".equals(targetName) && !"Dart".equals(targetName);
+			return !"".equals(targetName) && !"Swift".equals(targetName) && !"Dart".equals(targetName);
```
Why `Java` is removed here and below?

parrt replied:
hmm...good question. must be a typo or something weird when I was editing. hmm..yeah, then i saw that as always ignoring. It was failing java. let me add descr back and see.

parrt replied:
ok, descriptor seems to work. I added

```
[type]
Parser

[grammar]
grammar T;

member : 'a';
body : member*;
file : body EOF;
B : 'b';

[start]
file

[input]
baa

[output]

[errors]
"""line 1:0 mismatched input 'b' expecting {<EOF>, 'a'}
"""

[skip]
Cpp
CSharp
Go
Node
PHP
Python2
Python3
```

KvanTTT commented on runtime-testsuite/resources/org/antlr/v4/test/runtime/new_descriptors/CompositeParsers/BringInLiteralsFromDelegate.txt:
```diff
@@ -0,0 +1,23 @@
+=a
+
+[output]
+"""S.a
+"""
```
Is it possible to remove useless `"""` for almost all `output`? It looks untidy. It should be whitespace tolerant because it contains an output parse tree. The same for `errors`.

parrt replied:
For output newlines really matter because we might be generating that rather than simply calling println. For errors, it's true we can remove except then we'd have different rules for different sections of the file.

KvanTTT commented on runtime-testsuite/resources/org/antlr/v4/test/runtime/new_descriptors/CompositeLexers/LexerDelegatorInvokesDelegateRule.txt:
```diff
@@ -0,0 +1,24 @@
+WS : (' '|'\n') -> skip ;
+
+[slaveGrammar]
+		lexer grammar S;
+		A : 'a' {<writeln("\"S.A\"")>};
+		C : 'c' ;
```
Looks like incorrect indentation here.

KvanTTT commented on runtime-testsuite/resources/org/antlr/v4/test/runtime/new_descriptors/FullContextParsing/AmbigYieldsCtxSensitiveDFA.txt:
```diff
@@ -0,0 +1,27 @@
+s0-ID->:s1^=>1
+
+[errors]
+"""line 1:0 reportAttemptingFullContext d=0 (s), input='abc'
+"""
```
`"""` can be removed for all `[errors]`. It does not matter how many line breaks in errors, they can be trimmed.

parrt replied:
It's tricky with the new lines because sometimes we want to ignore white space and sometimes we don't. Probably for error messages you are right but let's start with a general translation and then figure out how to make it simpler. Although it might not be necessary since we don't really create that many new tests./

parrt commented:
Heh, @ericvergnaud can you take a peek? I think it works modulo random build failures that seem to be CI related.  This splits C++ tests to get them to avoid timeout. I tried to update the doc a bit

https://github.com/parrt/antlr4/blob/runtime-test-descriptor-files/doc/antlr-project-testing.md

Also i've left in the migration code so we don't lose (if branch is destroyed) in case we need it again.

parrt commented:
BTW, i'm unable to get anything but 1.8 src/trg to work even when compiling with jdk 11. Maven complains the target isn't compatible. booo! so, it's 1.8 -> 1.8 at moment; lemme know if you know how to make this work. I've tried old style and new `<release>1.8</release>` type stuff in maven.

KvanTTT commented on doc/antlr-project-testing.md:
````diff
@@ -190,68 +149,55 @@ Use this to run tests in parallel:
 ```bash
 $ export MAVEN_OPTS="-Xmx1G"
-$ mvn -Dparallel=methods -DthreadCount=4 test
+$ mvn -Dparallel=classes -DthreadCount=4 test
````
Why classes? It looks like all methods are isolated and they can be and should be tested in parallel. Also, the change is unrelated to the header of the pull request, I recommend dropping this change.

parrt replied:
I ran into a problem on my M1 Mac for Go; it was creating hundreds of swift-frontend threads and I almost had to reboot the system because I couldn't even get a prompt...

KvanTTT replied:
It's weird because there is a limit on threads count (4). Have switching to `classes` resolved the problem completely?

KvanTTT commented on doc/antlr-project-testing.md:
```diff
@@ -6,52 +6,56 @@ Because ANTLR supports multiple target languages, the unit tests are broken into
 
 ## Requirements
 
+*out of date, at least for mono*
```
Unrelated changes to the pull request. I suggest updating the requirements info in the next PR (replace mono with dotnet).

parrt replied:
just updating doc as part of broadened PR.

KvanTTT commented on runtime-testsuite/resources/org/antlr/v4/test/runtime/descriptors/CompositeLexers/LexerDelegatorRuleOverridesDelegate.txt:
```diff
@@ -0,0 +1,22 @@
+WS : (' '|'\n') -> skip ;
+
+[slaveGrammar]
+		lexer grammar S;
+		A : 'a' {<writeln("\"S.A\"")>} ;
+		B : 'b' {<writeln("\"S.B\"")>} ;
```
Useless indent.

KvanTTT commented on runtime/Go/antlr/utils.go:
```diff
@@ -8,7 +8,7 @@ import (
 	"errors"
 	"fmt"
-	"math/bits"
+	"sort"
```
Unrelated changes. Please remove them from PR (I recommend using a very convenient interactive rebase in IntelliJ for changing local history).

parrt replied:
that it's just a change when I reverted the Go done by @jcking in this branch when I was playing around.

KvanTTT reviewed (changes requested):
What's the meaning of keeping old descriptors in `OLD-descriptors` directory? I suggest removing them. Git keeps history very well.

Otherwise looks good! 👍🏼

KvanTTT commented:
> Heh, @ericvergnaud can you take a peek? I think it works modulo random build failures that seem to be CI related. This splits C++ tests to get them to avoid timeout. I tried to update the doc a bit
> 
> https://github.com/parrt/antlr4/blob/runtime-test-descriptor-files/doc/antlr-project-testing.md
> 
> Also i've left in the migration code so we don't lose (if branch is destroyed) in case we need it again.

I think it's also an unrelated change. Let's set up CI in another pull request. Moreover, it looks like there are some problems with GitHub CI because of [the following error](https://github.com/antlr/antlr4/runs/4588372189?check_suite_focus=true):

```
Can't find any online and idle self-hosted runner in the current repository, account/organization that matches the required labels: 'self-hosted , macOS , x64'
Waiting for a self-hosted runner to pickup this job...
```

parrt commented:
> What's the meaning of keeping old descriptors in `OLD-descriptors` directory? I suggest removing them. Git keeps history very well.

Yep, you're right. I have removed them.

parrt commented:
> I think it's also an unrelated change. Let's set up CI in another pull request. 

I have broadened the scope of the PR to be a reorganization of the entire runtime test rig; changed the title.

>Moreover, it looks like there are some problems with GitHub CI because of [the following error](https://github.com/antlr/antlr4/runs/4588372189?check_suite_focus=true):
> 
> ```
> Can't find any online and idle self-hosted runner in the current repository, account/organization that matches the required labels: 'self-hosted , macOS , x64'
> Waiting for a self-hosted runner to pickup this job...
> ```

Yeah, I don't know what the issue is there but I don't think it has something in the configuration I changed. Github is likely just pissed that we have so many actions coming in.

KvanTTT commented:
> Yeah, I don't know what the issue is there but I don't think it has something in the configuration I changed. Github is likely just pissed that we have so many actions coming in.

There is something wrong with @ericvergnaud machine because GitHub CI uses `self-hosted` configuration, see https://github.com/antlr/antlr4/pull/3425 thread. Changing `runs-on: [self-hosted, macOS, x64]` to `runs-on: macos-latest` resolved the problem, but probably not completely.

parrt commented:
Ok, i'm running some intellij code Analysis and cleaning up accordingly at moment then I think we can merge.  Next step, maybe somebody smarter than me can figure out how to get this to compile with 11 but generate 7 (or 8) haha

KvanTTT commented:
Looks like `self-hosted` GitHub CI is working again: https://github.com/antlr/antlr4/actions

parrt commented:
Let's see if it works for this PR. :)

parrt commented:
Looks like all unit tests pass except for swift-recursion due to timeout (works on my machine).  I think it's time to merge!!

KvanTTT commented:
> Looks like all unit tests pass except for swift-recursion due to timeout (works on my machine). I think it's time to merge!!

Mavens says it's not timeout but incorrect test: https://github.com/antlr/antlr4/runs/4612727990?check_suite_focus=true#step:6:542

parrt commented:
Yes but if you look at this `Time elapsed: 1,173.605` it appears to be a time out that's simply didn't get any output causing a field test. That is almost exactly 20 minutes.

KvanTTT commented:
It's the elapsed time for the entire test group and it's not so much. For instance, here we have 27 min and green tests (also RECURSION): https://github.com/antlr/antlr4/runs/4609645043?check_suite_focus=true

parrt commented:
Hmm... Interesting. OK well it passes elsewhere so I'm not sure why would fail over there. Let me see if it fails here locally on an identical box to @ericvergnaud's.

parrt commented:
All TestLeftRecursion tests pass on my identical M1 mac to the test box. So, the only difference would be the number of threads as I keep it single thread it here. That's good mean there's an issue with Swift testing and parallel.

ericvergnaud commented:
My experience is that GitHub runners do not behave consistently. The exact same commit will sometimes pass the tests in 5ms, other times in 10 mns, and other times timeout. This is not specific to Swift.
This also happens with other CIs, but more frequently with GitHub, possibly because we're using self-hosted runners, and the network conversation becomes less reliable.
Here's a ping trace from my box:
PING github.com (140.82.121.3): 56 data bytes
64 bytes from 140.82.121.3: icmp_seq=0 ttl=51 time=34.212 ms
64 bytes from 140.82.121.3: icmp_seq=1 ttl=51 time=52.130 ms
64 bytes from 140.82.121.3: icmp_seq=2 ttl=51 time=91.085 ms
64 bytes from 140.82.121.3: icmp_seq=3 ttl=51 time=71.296 ms
64 bytes from 140.82.121.3: icmp_seq=4 ttl=51 time=154.234 ms
64 bytes from 140.82.121.3: icmp_seq=5 ttl=51 time=119.809 ms
64 bytes from 140.82.121.3: icmp_seq=6 ttl=51 time=200.065 ms
64 bytes from 140.82.121.3: icmp_seq=7 ttl=51 time=184.354 ms
Request timeout for icmp_seq 8
64 bytes from 140.82.121.3: icmp_seq=9 ttl=51 time=148.027 ms
64 bytes from 140.82.121.3: icmp_seq=10 ttl=51 time=162.412 ms
64 bytes from 140.82.121.3: icmp_seq=11 ttl=51 time=67.909 ms
64 bytes from 140.82.121.3: icmp_seq=12 ttl=51 time=117.240 ms
64 bytes from 140.82.121.3: icmp_seq=13 ttl=51 time=136.301 ms
64 bytes from 140.82.121.3: icmp_seq=14 ttl=51 time=177.255 ms
64 bytes from 140.82.121.3: icmp_seq=15 ttl=51 time=108.688 ms
64 bytes from 140.82.121.3: icmp_seq=16 ttl=51 time=126.442 ms
Request timeout for icmp_seq 17
64 bytes from 140.82.121.3: icmp_seq=18 ttl=51 time=103.139 ms
64 bytes from 140.82.121.3: icmp_seq=19 ttl=51 time=151.846 ms
64 bytes from 140.82.121.3: icmp_seq=20 ttl=51 time=187.243 ms
^C
--- github.com ping statistics ---
21 packets transmitted, 19 packets received, 9.5% packet loss
round-trip min/avg/max/stddev = 34.212/125.984/200.065/46.485 ms

parrt commented:
Very interesting. So even if we bought a bigger mac or hosted at Amazon, we could still have problems communicating with GitHub

KvanTTT commented:
> My experience is that GitHub runners do not behave consistently. The exact same commit will sometimes pass the tests in 5ms, other times in 10 mns, and other times timeout. This is not specific to Swift.

Have you checked if it depends on simultaneous running pipelines? For instance, there are 6 running pipelines, it's many: https://github.com/antlr/antlr4/actions

ericvergnaud commented:
These 6 pipelines are queued, not running. We have 4 runners so can only run 4 pipelines simultaneously. Initially I implemented 8, but that turned out to be disastrous, so sacred down to 4.

KvanTTT commented:
What if scale down to 1 and try several runs?

ericvergnaud commented:
Not sure what that means tbh

parrt commented:
Wow. swift-build, driver, etc... all running in Intel mode on my M1. I'm playing with threads.

KvanTTT commented:
FYI, some runtime tests were lost during transformation, for instance `SemPredEvalParser/PredFromAltTestedInLoopBack_1`. But I've restored it in [my rebased PR](https://github.com/antlr/antlr4/pull/3349/files#diff-ae6a41c28cc4a52b8c7c0cdd918da6ca36b2e5195e44c7c13b73e77f27c93e40) (and fixed).

parrt commented:
Now I'm running with methods not classes:

```
mvn -Dparallel=methods -DthreadCount=4 -Dtest=swift.** test
```

Almost nothing being done in parallel. not sure why.

parrt commented:
> FYI, some runtime tests were lost during transformation, for instance `SemPredEvalParser/PredFromAltTestedInLoopBack_1`. 

was worried about that. i have a todo on my list to count :) but thanks

KvanTTT commented:
Maybe it related to some bug with test inheritance (PredFromAltTestedInLoopBack_1 and PredFromAltTestedInLoopBack_2 had common base).

parrt commented:
yeah, i saw a few that didn't get translated but was usually a missed ignored() result or something. could be inheritance yep.

KvanTTT commented:
> I've seen 40 builds simultaneously.

Yes, looks weird, most part of them should be killed.

parrt commented:
Ok, `mvn -Dparallel=methods -DthreadCount=4 -Dtest=swift.** test` does nothing in parallel that I can see:

<img width="858" alt="Screen Shot 2021-12-23 at 2 20 44 PM" src="https://user-images.githubusercontent.com/178777/147296048-bd56ee97-3d5c-48d9-aa3a-529f55044da2.png">
